packs/mssql/actions/lib/results_processor.py

```diff
@@ -0,0 +1,104 @@
+
+    def __init__(self, config, logger):
+        self.config = config
+        self.logger = logger
+
+    def execute_scalar(self, response, cursor):
+        """
+        Returns the scalar response.
+        """
+        if not response:
+            sys.exit(self.NO_DATA)
+        return response
+
+    def execute_row(self, response, cursor):
+        """
+        Returns a mapping of column name to value for a row.
+        """
+        row = self._filter_numbered_columns(response or {})
+        if not row:
+            sys.exit(self.NO_DATA)
+        return row
+
+    def execute_insert(self, response, cursor):
+        """
+        Returns identity value of the inserted row.
+        """
+        return cursor.identity
+
+    def execute_non_query(self, response, cursor):
+        """
+        Returns the number of rows affected by the non-query.
+        """
+        if cursor.rows_affected == 0:
+            sys.exit(self.NO_DATA)
+        return cursor.rows_affected
+
+    def execute_query(self, response, cursor):
+        """
+        Writes results to CSV for downstream processing. Each result set is written to its own file.
+        Returns a list of all file names in order of result set for use by downstream actions.
+
+        Checks `output_csv` section in `config.yaml` to determine where to write CSV files. You can
+        specify the output `directory` as well as the file `prefix` and `suffix`.
+
+        Tries writing to `$TMPDIR`, `$TEMP`, and `$TMP` in order before falling back to platform-
+        specific locations. See https://docs.python.org/2/library/tempfile.html#tempfile.tempdir.
+        """
+        output_files = []
+        while True:
+            with self._get_output_file() as csv_file:
+                # Since pythonrunner runs as root, only root can read and write CSV file by default
+                # Let's chmod it so downstream processes which run as stanley can read and write
+                os.chmod(csv_file.name, 0o666)  # race condition w/ open() OK since increasing perms
+                try:
+                    # pylint: disable=no-member
+                    # Grab the first row so we can read the headers and write them to the CSV
+                    first_row = self._filter_numbered_columns(
+                        next(cursor.get_iterator(_mssql.ROW_FORMAT_DICT)))
+                except StopIteration:
+                    # the last result set will always be empty, so remove the file created for it
+                    os.unlink(csv_file.name)
+                    # we're done, no more result sets
+                    break
+                output_files.append(csv_file.name)
+                writer = csv.DictWriter(csv_file, fieldnames=first_row.keys())
+                writer.writeheader()
+                writer.writerow(first_row)
+                for row in cursor:
+                    writer.writerow(self._filter_numbered_columns(row))
+        if not output_files:
```

Comment:
Yeah, I was hoping someone would bring this usage point up. The current is modeled after `grep`, which returns a failure exit code when it doesn't find a match. My original thinking was that it would make decisions in workflows easier.

For example, in the `execute.query_and_email` action, we specifically don't want to email if there's `NO_DATA` returned. Of course, we don't want this to show up as a failure though, so we end up adding an extra task in the workflow anyway. 

```yaml
  -
    name: 'query_mssql'
    ref: 'mssql.execute.query'
    parameters:
      query_string: '{{query_string}}'
      params: '{{params}}'
      database: '{{database}}'
    # success indicates returned data. don't email if NO DATA
    on-success: 'email_results'
    # failure may indicate no data or an actual error
    on-failure: 'success_on_no_data'
 ```

I guess we either go from on-success into a "router" action which decides the next action based on whether it `HAS_DATA` or `NO_DATA`, or we go from on-failure into a similar action which decides based on `NO_DATA` or `HAS_ERROR`. In other words, we have three possible states but only two outputs, so either way isn't super clean.

Do you have any better thoughts, suggestions, or patterns here? For example, should failure always indicate an exception, or can it be used for conveying another meaning (e.g., `NO_DATA`)?